### PR TITLE
Check for assignments between non-conflicting generator saved locals

### PR DESCRIPTION
**Pull request by ecstatic-morse**
This is to prevent future changes to the generator transform from reintroducing the problem that caused #73137. Namely, a store between two generator saved locals whose storage does not conflict.

My ultimate goal is to introduce a modified version of #71956 that handles this case properly.

r? @tmandry

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
I ran this locally on the `src/test/ui/{async-await,generator}` tests with `if opts.validate_mir` part removed and everything passed.

File: src/librustc_mir/transform/generator.rs [hunk 1]
```diff
@@ -1314,3 +1340,130 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
         create_generator_resume_function(tcx, transform, source, body, can_return);
     }
 }
+
+/// Looks for any assignments between locals (e.g., `_4 = _5`) that will both be converted to fields
+/// in the generator state machine but whose storage does not conflict.
```
Comment (tmandry) [inline src/librustc_mir/transform/generator.rs:1345]: At first I was confused by the wording here. To me the storage _does_ conflict, we just aren't aware of it. Maybe:

```suggestion
/// in the generator state machine but whose storage is not marked as conflicting.
```

File: src/librustc_mir/transform/generator.rs [hunk 2]
```diff
@@ -1314,3 +1340,130 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
...
+        if place.is_indirect() {
+            return None;
+        }
+
+        self.saved_locals.get(place.local)
+    }
+
+    fn check_assigned_place(&mut self, place: Place<'tcx>, f: impl FnOnce(&mut Self)) {
+        if let Some(assigned_local) = self.saved_local_for_direct_place(place) {
+            assert!(self.assigned_local.is_none(), "`check_assigned_local` must not recurse");
```
Comment (tmandry) [inline src/librustc_mir/transform/generator.rs:1373]: nit:
```suggestion
            assert!(self.assigned_local.is_none(), "`check_assigned_place` must not recurse");
```

File: src/librustc_mir/transform/generator.rs [hunk 3]
```diff
@@ -1314,3 +1340,130 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
...
+        }
+    }
+}
+
+impl Visitor<'tcx> for EnsureGeneratorFieldAssignmentsNeverAlias<'_> {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+        let lhs = match self.assigned_local {
+            Some(l) => l,
+            None => {
+                // We should be visiting all places used in the MIR.
```
Comment (tmandry) [inline src/librustc_mir/transform/generator.rs:1387]: I think the wording of this comment could be improved. The correctness of the code seems to hinge on the fact that we skip running `f` in `check_assigned_place` if the lhs is not saved. In those cases we don't visit all places used in the MIR.

Maybe the more generic "handling" is better?

```suggestion
                // We should be handling all place uses.
```

**Review by tmandry — APPROVED** [review]
Looks good! r=me after nits.

File: src/librustc_mir/transform/generator.rs [hunk 4]
```diff
@@ -1315,3 +1341,134 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
...
+                        this.visit_operand(arg, location);
+                    }
+                });
+            }
+
+            TerminatorKind::Yield { value, resume: _, resume_arg, drop: _ } => {
+                self.check_assigned_place(*resume_arg, |this| this.visit_operand(value, location));
+            }
+
+            // FIXME: Does `asm!` have any aliasing requirements?
```
Comment (bjorn3) [inline src/librustc_mir/transform/generator.rs:1456]: I don't think so. All inputs are completely moved to registers before running the inline asm and storing the outputs again.

File: src/librustc_mir/transform/generator.rs [hunk 5]
```diff
@@ -1315,3 +1341,134 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
         create_generator_resume_function(tcx, transform, source, body, can_return);
     }
 }
+
+/// Looks for any assignments between locals (e.g., `_4 = _5`) that will both be converted to fields
```
Comment (Aaron1011) [inline src/librustc_mir/transform/generator.rs:1345]: Why do we only need to validate assignments where the RHS is a direct usage of a local? Can indirect usages not cause this issue?

Comment (ecstatic-morse) [reply src/librustc_mir/transform/generator.rs:1345]: Good point.

Is MIR such as `*p = *q` or `*p = q` valid when `p` overlaps with `q`? `p = *q` is almost certainly invalid, and should be handled by this check. I'll check for all three versions in a subsequent PR.

**Pull request merged**